### PR TITLE
Adds the Active Sonar mod to the game. 

### PR DESCRIPTION
**Pull request by Wallemations**
## About The Pull Request
Adds the Active Sonar Module to the game, a module which lets you see the locations of living creatures within a 9 tile radius.
It can be attained by researching Security Modules, and then printed like any other module. 
It takes 3 complexity to house, has a 25 second cooldown, and takes a good amount of energy to use.

https://user-images.githubusercontent.com/66052067/174206204-6218ede6-26d4-48c4-ad25-1566a6472e67.mp4

(The cooldown for the video above has been lowered for demonstration purposes)

The sonar sound is a mix of two sounds from freesound.org;
[Hit](https://freesound.org/people/pushtobreak/sounds/16793/)
[Sonar](https://freesound.org/people/marb7e/sounds/620321/)

## Why It's Good For The Game
This is essentially a downgraded version of Thermal Vision from Genetics, leaving the player without the eye damage, but being far less accurate overall. The main benefit of including this in the game is the unique interaction between chaser and suspect, that being that anyone within the ping radius will be able to hear the loud ping created from the sonar. This gives the suspect a small window of time to try and escape before they're caught, heightening tension. Additionally, if someone wanted to try and hide from the chaser, they could gather and scatter other living mobs, which could divert a chaser long enough for the suspect to slip away. Overall, I think it's a far more interesting back-and-forth system than thermal vision.

## Changelog

:cl: Wallem
add: Adds the Active Sonar Module, a module that allows the suit-wearer to detect living organisms within a given radius.
/:cl:

### CODE REVIEW

**Comment by Shadowflame909** [comment]
This + X-ray laser guns could be a very unique way of getting that murderbone on

**Comment by MTandi** [comment]
Wait, you can make module icons appear directly on the bar and not in the circle?

Also, how does it work with moving targets? Is the circle attached to the entity?

**Comment by necromanceranne** [comment]
Nice.

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 1]
```diff
@@ -543,3 +543,36 @@
...
+
+/obj/effect/temp_visual/sonar_ping/Initialize(mapload, mob/living/looker, mob/living/creature)
+	. = ..()
+	modsuit_image = image(icon = icon, loc = looker.loc, icon_state = real_icon_state, layer = ABOVE_ALL_MOB_LAYER, pixel_x = ((creature.x - looker.x) * 32), pixel_y = ((creature.y - looker.y) * 32))
+	modsuit_image.plane = ABOVE_LIGHTING_PLANE
+	mod_man = WEAKREF(looker)
+	add_mind(looker)
+
+/obj/effect/temp_visual/sonar_ping/Destroy()
+	remove_mind(mod_man.resolve())
```
Comment (Fikou) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:567]: what if mod_man doesn't exist?

Comment (Fikou) [reply code/game/objects/effects/temporary_visuals/miscellaneous.dm:567]: i suggest you do something like
var/mob/living/previous_user = mod_man.resolve()
if(previous_user)
remove_mind(previous_user)

Comment (Wallemations) [reply code/game/objects/effects/temporary_visuals/miscellaneous.dm:567]: Sure thing, I'll do it when I get home

**Comment by Wallemations** [comment]
so I have zero clue what to do to fix this map integration test failure
```
REASON #5: /obj/effect/temp_visual/sonar_ping failed to return a qdel hint 1 times out of a total del count of 4020 at code/modules/unit_tests/create_and_destroy.dm:167
```

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 2]
```diff
@@ -543,3 +543,38 @@
...
+	randomdir = FALSE
+	/// The image shown to modsuit users
+	var/image/modsuit_image
+	/// The person in the modsuit at the moment, really just used to remove this from their screen
+	var/datum/weakref/mod_man
+	/// The icon state applied to the image created for this ping.
+	var/real_icon_state = "sonar_ping"
+
+/obj/effect/temp_visual/sonar_ping/Initialize(mapload, mob/living/looker, mob/living/creature)
+	. = ..()
```
Comment (Fikou) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:560]: you want to check here if looker and creature exist, if not return a qdel initialize hint

**Comment by Fikou** [comment]
> Wait, you can make module icons appear directly on the bar and not in the circle?

my brother in christ you have been able to for the last 5 months

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 3]
```diff
@@ -543,3 +543,40 @@
...
+	if(!looker || !creature)
+		return INITIALIZE_HINT_QDEL
+	modsuit_image = image(icon = icon, loc = looker.loc, icon_state = real_icon_state, layer = ABOVE_ALL_MOB_LAYER, pixel_x = ((creature.x - looker.x) * 32), pixel_y = ((creature.y - looker.y) * 32))
+	modsuit_image.plane = ABOVE_LIGHTING_PLANE
+	mod_man = WEAKREF(looker)
+	add_mind(looker)
+
+/obj/effect/temp_visual/sonar_ping/Destroy()
+	var/mob/living/previous_user = mod_man.resolve()
```
Comment (Fikou) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:569]:
```suggestion
	var/mob/living/previous_user = mod_man?.resolve()
```

**Review by tralezab — APPROVED** [review]
wow that is some clean code

**Pull request merged**